### PR TITLE
Tests and bugfixes

### PR DESCRIPTION
**Pull request by Clydingus**
Implements more flexible softmax-n param implementations, now can use different floats for the denominator. Removes `inspect_softmax_sum` flag in favour of simply calling method `compute_softmax_metrics(self)`. 

Note:
```
        if self.flash and args.softmax1:
            print("softmax1 is not supported in flash attention, not using flash attention")
            self.flash = False
```
is actually a duplicate fix for flashattention bug that is already present in master branch, should be removed in favour of existing master branch fix.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -471,3 +494,39 @@ def generate(
             return idx, torch.cat(logits_list, dim=1)
 
         return idx
+
+    # *yoink*: https://github.com/tcapelle/llama2.c/blob/af8597ca6ac0c74142951c44b8d6926e170d791a/model.py#L428
+    # the following metric computations require that the model be run in eval mode, then the respective layer outputs
+    # will be saved in order to compute the metric of the latest input passed through the model
+    def compute_attention_metrics(self) -> Tuple[List[float], List[float]]:
```
Comment (photomz) [inline model.py:501]: Turn these compute_ funcs into `@classmethod`s

File: train.py [hunk 2]
```diff
@@ -36,7 +36,7 @@
 
 # -----------------------------------------------------------------------------
 # I/O
-out_dir = "out"
+out_dir = "out/softmax1-110m"
```
Comment (photomz) [inline train.py]: 110m??

File: train.py [hunk 3]
```diff
@@ -48,21 +48,21 @@
...
 vocab_source = (
     "llama2"  # llama2|custom; use Lllama 2 vocab from Meta, or custom trained
 )
 vocab_size = 32000  # the Llama 2 tokenizer has 32K tokens
 # model
-dim = 288
-n_layers = 6
-n_heads = 6
-n_kv_heads = 6
+dim = 512
```
Comment (photomz) [inline train.py]: Why? Can revert?

File: train.py
Comment (photomz) [inline train.py:1]: Revert most of these changes

**Review by photomz — COMMENTED** [review]
LGTM! Small cleanups

File: model.py (hunk 1 above)
Comment (photomz) [reply model.py:501]: Nvm LGTM

File: train.py [hunk 4]
```diff
@@ -50,19 +50,20 @@
...
+n_layers = 8
+n_heads = 8
+n_kv_heads = 8
```
Comment (photomz) [inline train.py]: These changes still aren't reverted -- check more pls

**Review by photomz — APPROVED** [review]
Let's goooo 🚀
